### ISSUE TITLE
upload-js-full SDK throws on import: Uncaught TypeError: Object.defineProperty called on non-object 

### ISSUE THREAD

**Issue by mbrookson** [upload-js/upload-js#9]
I'm trying to use the `upload-js-full` SDK in a Next.js 13 application. (`"next": "^13.4.2",`)

I'm seeing the following error:

```
Uncaught TypeError: Object.defineProperty called on non-object
    at Function.defineProperty (<anonymous>)
    at __webpack_require__.r (webpack.js?ts=1686572445609:205:21)
    at eval (main.mjs:1:21)
    at ./node_modules/.pnpm/upload-js-full@2.14.0/node_modules/upload-js-full/dist/browser/esm/main.mjs (manage.js?ts=1686572445609:46870:1)
    at options.factory (webpack.js?ts=1686572445609:661:31)
    at __webpack_require__ (webpack.js?ts=1686572445609:37:33)
    at fn (webpack.js?ts=1686572445609:316:21)
    at eval (upload.ts:8:72)
    at ./src/utils/upload.ts (manage.js?ts=1686572445609:46689:1)
    at options.factory (webpack.js?ts=1686572445609:661:31)
    at __webpack_require__ (webpack.js?ts=1686572445609:37:33)
    at fn (webpack.js?ts=1686572445609:316:21)
    at eval (ManageListing.tsx:18:71)
    at ./src/components/pages/owners/listings/ManageListing.tsx (manage.js?ts=1686572445609:46645:1)
    at options.factory (webpack.js?ts=1686572445609:661:31)
    at __webpack_require__ (webpack.js?ts=1686572445609:37:33)
    at fn (webpack.js?ts=1686572445609:316:21)
    at eval (manage.tsx:8:105)
    at ./src/pages/owners/listings/[id]/manage.tsx (manage.js?ts=1686572445609:46678:1)
    at options.factory (webpack.js?ts=1686572445609:661:31)
    at __webpack_require__ (webpack.js?ts=1686572445609:37:33)
    at fn (webpack.js?ts=1686572445609:316:21)
    at eval (?d8e5:5:16)
    at eval (route-loader.js:230:51)
```
I may be wrong but it appears to be we webpack error that's happening at the point of importing the dependency. This is a complete blocker for using the SDK.

Any ideas what is causing the issue?

For reference, the minimal `upload-js` SDK does work as expected, however I need the full SDK as I want access to more SDK actions and want use it on both the client and server.

**Comment by ljwagerfield** [comment]
Thank you for raising!

I'm actually struggling to get `import * as Upload from 'upload-js-full';` to work with Next.js 13: works in vanilla Node.js but for some reason, Next.js doesn't see any of the exports.

While we're looking into this, please can you create a small reproduction repo w/ README to demonstrate the problem you're having?

**Comment by mbrookson** [comment]
Thanks for the response!

Just been playing around with some fresh projects. The error I'm seeing is in a project that was initialised from the T3 stack (https://create.t3.gg/).

However, I've just set up a completely fresh Next.js 13 app, AND a fresh T3 app and it got it to work immediately with a really simple example, which is good I suppose!

This `index.tsx` page works with a completely empty fresh project:

```tsx
import Head from "next/head";
import { Configuration, UploadManager } from "upload-js-full";

const uploadManager = new UploadManager(
  new Configuration({
    fetchApi: fetch,
    apiKey: "XXXX", // e.g. "secret_xxxxx"
  })
);

export default function Home() {
  const upload = async (file: File) => {
    await uploadManager.upload({
      accountId: "XXXX",
      data: file,
    });
  };

  const onFileChange = async (e: React.ChangeEvent<HTMLInputElement>) => {
    if (!e.target.files) return;

    const file = e.target.files[0];

    if (!file) return;

    await upload(file);
  };

  return (
    <>
      <Head>
        <title>Upload.io PoC</title>
        <meta name="description" content="Generated by create next app" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <link rel="icon" href="/favicon.ico" />
      </Head>
      <main>
        <input type="file" onChange={(e) => void onFileChange(e)} />
      </main>
    </>
  );
}

```

Therefore even if there is a problem with the `upload-js-full` library it seems like an edge case. 

Perhaps some other dependency my project is interfering somehow. I'll do some more investigation and trial/error and report back.

Any further ideas welcome. I'll provide an example repo if I can replicate it.

**Comment by ljwagerfield** [comment]
Thank you for taking the time to share this: really appreciated!

Also glad you've got it working!

We're still unable to get a basic `import` working with Next.js 13 on node v16.19.0.

Which version of Node are you using, out of curiosity?

**Comment by mbrookson** [comment]
Aha! Problem solved for me!

I noticed a difference is that I'm using `pnpm` as my package manager in the project with the error. I just created a fresh repo again but this time using `pnpm` and voila, it causes the error!

From some more testing it seems it was an issue with using a combination of Next.js and pnpm together which has since been fixed in a patch version update to Next.js. Had a quick look through some of the changelog (there's a lot!) but not entirely sure what the actual fix was.

Runtime:
```
node: 18.12.1
pnpm: 8.6.0
```

`next: 13.4.2` ❌  Fails with `Uncaught TypeError: Object.defineProperty called on non-object`
`next: 13.4.5` ✅  Succeeds! 

Sorry for the trouble 😅 

Want me to push up a simple working Next.js demo for you still?

**Comment by ljwagerfield** [comment]
> Want me to push up a simple working Next.js demo for you still?

Only if it's not too much trouble!

Next.js is obviously a very popular framework, and it's a little troubling for us that the internal demo project we created as part of resolving this ticket doesn't even compile 😅 

From what we can tell, our ESM modules look valid and work fine with vanilla Node.js -- in any case -- it'd be a massive help to have your repo to test against 😄

**Comment by mbrookson** [comment]
No problem. Here's a link: https://github.com/mbrookson/upload-io-poc

Good luck! 🙌

**Comment by ljwagerfield** [comment]
This is very helpful, thank you!

It works for us too, btw, which is interesting. 

I'll leave this ticket open to track [this issue](https://github.com/upload-io/upload-js/issues/9#issuecomment-1588003142) which can be reproduced by:

```
npx create-next-app@latest
cd my-app
npm install upload-js-full node-fetch
```

and then adding this to the top of `src/app/page.tsx`:

```typescript
import * as Upload from "upload-js-full";
import fetch from "node-fetch"; // Node.js only.

const uploadManager = new Upload.UploadManager(
  new Upload.Configuration({
    fetchApi: fetch,
    apiKey: "YOUR_UPLOAD_API_KEY" // e.g. "secret_xxxxx"
  })
);
```

and then

```
npm run dev
```

error:

```
Attempted import error: 'UploadManager' is not exported from 'upload-js-full' (imported as 'Upload').

Import trace for requested module:
./src/app/page.tsx

./src/app/page.tsx
Attempted import error: 'Configuration' is not exported from 'upload-js-full' (imported as 'Upload').

Import trace for requested module:
./src/app/page.tsx
```

**Comment by mbrookson** [comment]
One thing to note is my example uses the pages router rather than the new app router.

Maybe there's something different about imports with the new one?

**Comment by ljwagerfield** [comment]
We've since been able to reproduce this error.

It's now fixed in `upload-js-full@^2.18.0`

Thank you so much @mbrookson!
